source/s3_auto_ranged_put.c

```diff
@@ -174,6 +191,45 @@ static void s_s3_meta_request_auto_ranged_put_destroy(struct aws_s3_meta_request
     aws_mem_release(meta_request->allocator, auto_ranged_put);
 }
 
+static void s_s3_auto_ranged_put_cancel(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_request *failed_request,
+    int error_code) {
+    // TODO: error code needs to be propagated
+    (void)error_code;
+    struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
+
+    s_s3_auto_ranged_put_lock_synced_data(auto_ranged_put);
+    if (auto_ranged_put->synced_data.state != AWS_S3_AUTO_RANGED_PUT_STATE_CANCEL &&
+        auto_ranged_put->synced_data.state != AWS_S3_AUTO_RANGED_PUT_STATE_WAITING_FOR_CANCEL) {
+
+        if (auto_ranged_put->synced_data.num_parts_sent == auto_ranged_put->synced_data.num_parts_completed) {
+            /* No need to wait, no more parts will be sent. Safely abort the put */
+            auto_ranged_put->synced_data.state = AWS_S3_AUTO_RANGED_PUT_STATE_CANCEL;
+        } else {
+            /* In case there still put requests in flight, wait until they completed to switch to cancel state */
+            auto_ranged_put->synced_data.state = AWS_S3_AUTO_RANGED_PUT_STATE_WAITING_FOR_CANCEL;
+        }
+        auto_ranged_put->synced_data.failed_request = failed_request;
+    }
+    s_s3_auto_ranged_put_unlock_synced_data(auto_ranged_put);
+}
+
+static void s_s3_auto_ranged_put_notify_request_destroyed(
+    struct aws_s3_meta_request *meta_request,
+    struct aws_s3_request *request) {
+    struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
+    bool cancel = false;
+    s_s3_auto_ranged_put_lock_synced_data(auto_ranged_put);
+    cancel = request->request_tag == AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_ABORT_MULTIPART_UPLOAD;
+    s_s3_auto_ranged_put_unlock_synced_data(auto_ranged_put);
+    if (cancel) {
+        /* TODO: What's the error code here? */
+        aws_s3_meta_request_cancel_default(
+            meta_request, auto_ranged_put->synced_data.failed_request, AWS_ERROR_S3_CANCELED_SUCCESS);
+    }
```

Comment:
how to propagate the error happens from cancel or failed request to here? @rccarper